util/chunk/chunk.go

```diff
@@ -39,15 +44,57 @@ const (
 
 // NewChunkWithCapacity creates a new chunk with field types and capacity.
 func NewChunkWithCapacity(fields []*types.FieldType, cap int) *Chunk {
+	return New(fields, cap, cap) //FIX ME in continue PR, this keep every thing same.
+}
+
+// New creates a new chunk.
+//  cap: the max number of rows that the chunk can hold.
+//  maxChunkSize: the max limit for max number of rows.
+func New(fields []*types.FieldType, cap, maxChunkSize int) *Chunk {
 	chk := new(Chunk)
 	chk.columns = make([]*column, 0, len(fields))
-	chk.numVirtualRows = 0
 	for _, f := range fields {
-		chk.addColumnByFieldType(f, cap)
+		elemLen := getFixedLen(f)
+		if elemLen == varElemLen {
+			chk.columns = append(chk.columns, newVarLenColumn(cap, nil))
```

Comment:
we should firstly calculate the `capacity` according to the input parameters:
```go
chk.capacity = mathutil.Min(cap, maxChunkSize)
```

then use `chk.capacity` to create columns.